### DISCUSSION TITLE
Using a local usb webcam?

### DISCUSSION THREAD

**Discussion by DamonLane** [roflcoopter/viseron#705] [Q&A]
Hello, I appreciate the clear documentation but didn't see this possibility mentioned. I'm setting up a single webcam connected to a low powered ARM device, mainly because both are just laying around. I can access the webcam with ffmpeg from inside the Viseron container, but if I skip the host and port inputs and only provide the local path to the camera, Viseron errors:
> 
> Failed setup
> Error validating config for domain camera and component ffmpeg: required key not provided @ data['camera_1']['port']. Got None

That's not surprising, but is there a way to communicate to Viseron what I'm trying to do? The Gstreamer component gives the same message.

I could use ffserver or something similar to provide the steam on localhost where it would have the parameters the camera components expect, but if there's a way to configure the usb source directly it would be more efficient and simpler.

**Comment by bsyomov** [comment]
Also, you can use [ffmpeg component raw_command](https://viseron.netlify.app/components-explorer/components/ffmpeg#raw-command) to use any ffmpeg compatible source.

**Reply by DamonLane** [reply]
Thanks! That's promising, but even if I provide the usb input path to a ffmpeg raw_command, it seems like Viseron is checking for host and port inputs:
> 
>Error validating config for domain camera and component ffmpeg: required key not provided @ data['camera_one']['host']. Got None >required key not provided @ data['camera_one']['path']. Got None required key not provided @ data['camera_one']['port']. Got None
>
In the usb case, I do have a path, but not a host and port. In a network camera case, people would have those, but that info could be included in the raw_command. It looks like a Viseron check is my immediate obstacle and I can't get past that to see what Viseron is receiving from ffmpeg.

**Reply by bsyomov** [reply]
You can fill it with fake data, for config validation only. These options are not used with raw_command.

**Reply by DamonLane** [reply]
Thanks! I got past the config validation, but not much further. Now Viseron says:
`
FFprobe could not connect to stream. Output: {'error': {'code': -110, 'string': 'Connection timed out'}}`

I see that FFprobe is a ffmpeg command so maybe my troubleshooting has to move there. But I'll share the output I get from from running the raw_command in Viseron's docker, in case it reveals anything. Since I'm running a headless server, the illegible characters ffmpeg puts into the standard output is what is making me assume it's working and that's what Viseron is looking for but maybe that's not correct.

My Viseron config, with a real path just in case is:
```
ffmpeg:
  camera:
    camera_one:
      name: bird bath webcam
      host: 111.111.111.111
      port: 1111
      path: /dev/video6
      raw_command: ffmpeg -i /dev/video6 -vf fps=1.0 -f rawvideo -pix_fmt nv12 pipe:1
```
With Viseron running, I see FFprobe in top's process list, but not ffmpeg. Whereas I do see ffmpeg in top after running the command below. All that might mean is that Viseron closes the process when it fails whereas the naked command doesn't. (But why does it fail?) My test outside of Viseron is:
```
root@DietPi:/srv/viseron# docker exec viseron ffmpeg -i /dev/video6 -vf fps=1.0 -f rawvideo -pix_fmt nv12 cmd_docker_exec_ffmpeg
ffmpeg version 5.1.2 Copyright (c) 2000-2022 the FFmpeg developers
  built with gcc 9 (Ubuntu 9.4.0-1ubuntu1~20.04.1)
  configuration: --disable-debug --disable-doc --disable-ffplay --enable-shared --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-gpl --enable-libass --enable-fontconfig --enable-libfreetype --enable-libvidstab --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libxcb --enable-libx265 --enable-libxvid --enable-libx264 --enable-nonfree --enable-openssl --enable-libfdk_aac --enable-postproc --enable-small --enable-version3 --enable-libbluray --enable-libzmq --extra-libs=-ldl --prefix=/opt/ffmpeg --enable-libopenjpeg --enable-libkvazaar --enable-libaom --extra-libs=-lpthread --enable-neon --enable-v4l2_m2m --extra-cflags=-I/opt/ffmpeg/include --extra-ldflags=-L/opt/ffmpeg/lib
  libavutil      57. 28.100 / 57. 28.100
  libavcodec     59. 37.100 / 59. 37.100
  libavformat    59. 27.100 / 59. 27.100
  libavdevice    59.  7.100 / 59.  7.100
  libavfilter     8. 44.100 /  8. 44.100
  libswscale      6.  7.100 /  6.  7.100
  libswresample   4.  7.100 /  4.  7.100
  libpostproc    56.  6.100 / 56.  6.100
Input #0, video4linux2,v4l2, from '/dev/video6':
  Duration: N/A, start: 449.073575, bitrate: 147456 kb/s
  Stream #0:0: Video: rawvideo (YUY2 / 0x32595559), yuyv422, 1280x720, 147456 kb/s, 10 fps, 10 tbr, 1000k tbn
Stream mapping:
  Stream #0:0 -> #0:0 (rawvideo (native) -> rawvideo (native))
Press [q] to stop, [?] for help
Output #0, rawvideo, to 'cmd_docker_exec_ffmpeg':
  Metadata:
    encoder         : Lavf59.27.100
  Stream #0:0: Video: rawvideo (NV12 / 0x3231564E), nv12(tv, progressive), 1280x720, q=2-31, 11059 kb/s, 1 fps, 1 tbn
    Metadata:
      encoder         : Lavc59.37.100 rawvideo
frame=    3 fps=0.0 q=-0.0 size=    2560kB time=00:00:02.00 bitrate=10485.8kbits/s speed=3.2
frame=    3 fps=2.7 q=-0.0 size=    2560kB time=00:00:02.00 bitrate=10485.8kbits/s speed=1.7
frame=    4 fps=2.4 q=-0.0 size=    3840kB time=00:00:03.00 bitrate=10485.8kbits/s speed=1.8
frame=    4 fps=1.9 q=-0.0 size=    3840kB time=00:00:03.00 bitrate=10485.8kbits/s speed=1.3
frame=    5 fps=1.9 q=-0.0 size=    5376kB time=00:00:04.00 bitrate=11010.0kbits/s speed= 1.
frame=    5 fps=1.6 q=-0.0 size=    5376kB time=00:00:04.00 bitrate=11010.0kbits/s speed=1.2
frame=    6 fps=1.6 q=-0.0 size=    6656kB time=00:00:05.00 bitrate=10905.2kbits/s speed=1.3
frame=    6 fps=1.4 q=-0.0 size=    6656kB time=00:00:05.00 bitrate=10905.2kbits/s speed= 1.
frame=    7 fps=1.5 q=-0.0 size=    7936kB time=00:00:06.00 bitrate=10835.3kbits/s speed=1.2
frame=    7 fps=1.3 q=-0.0 size=    7936kB time=00:00:06.00 bitrate=10835.3kbits/s speed=1.1
frame=    8 fps=1.4 q=-0.0 size=    9216kB time=00:00:07.00 bitrate=10785.4kbits/s speed=1.2
frame=    8 fps=1.3 q=-0.0 size=    9216kB time=00:00:07.00 bitrate=10785.4kbits/s speed=1.1
frame=    9 fps=1.3 q=-0.0 size=   10752kB time=00:00:08.00 bitrate=11010.0kbits/s speed=1.1
frame=    9 fps=1.2 q=-0.0 size=   10752kB time=00:00:08.00 bitrate=11010.0kbits/s speed=1.1
frame=   10 fps=1.3 q=-0.0 size=   12032kB time=00:00:09.00 bitrate=10951.8kbits/s speed=1.1
frame=   10 fps=1.2 q=-0.0 size=   12032kB time=00:00:09.00 bitrate=10951.8kbits/s speed=1.0
```

**Reply by roflcoopter** [reply]
If you fill in width, height, fps, codec and audio_codec ffprobe should be skipped.

raw_command is quite new still and I frankly did not think of it being this complicated to use, sorry about that, will make sure it's easier in the future!

**Reply by roflcoopter** [reply]
Opened an issue to track it properly #708
